client/landing/gutenboarding/components/window-locale-effect-manager/index.tsx

```diff
@@ -0,0 +1,33 @@
+/**
+ * External dependencies
+ */
+import { switchWebpackCSS } from '../../../../lib/i18n-utils/switch-locale';
+import { useI18n } from '@automattic/react-i18n';
+import * as React from 'react';
+
+export const WindowLocaleEffectManager: React.FunctionComponent = () => {
+	const { __, isRTL, i18nLocale } = useI18n();
+
+	// Some languages may need to set an html lang attribute that is different from their slug
+	let lang = __( 'html_lang_attribute' );
+
+	// Some languages don't have the translation for html_lang_attribute
+	// or maybe we are dealing with the default `en` locale. Return the general purpose locale slug
+	if ( lang === 'html_lang_attribute' ) {
+		lang = i18nLocale;
+	}
+
+	React.useEffect( () => {
+		document.documentElement.lang = lang;
+	}, [ lang ] );
+
```

Comment:
```suggestion
	// Ensure `ltr` string will be extracted into calypso-strings.pot
	// Since string extraction is __ calls as if they were i18n-calypso's translate calls,
	// `context` must be passed the same way.
	__( 'ltr', { context: 'text direction' } );
```

I discovered that `isRTL` wasn't working with translation chunks and it's related to how [`@wordpress/i18n` implements it](https://github.com/WordPress/gutenberg/blob/master/packages/i18n/src/create-i18n.js#L185). Internally, it's calling gettext with `ltr` original string, but since it's not part of Calypso's monorepo, it ends up not being included in `calypso-strings.pot`, and therefore not being included in Gutenboarding's generated translation chunks. As a workaround, we can simply call the gettext to ensure `ltr` will be extracted as original string from the repo.

There's another catch here - we're are still using `i18n-calypso` to extract strings ([switching to alternative workflow is in the progress](https://github.com/Automattic/wp-calypso/pull/39541)) so `__` calls are actually being handled as if they were `translate` calls from `i18n-calypso`, and therefore `context` must be passed the same way.

Reply:
Thanks for the investigation and finding this bug!

hmmm I still don't understand this, I tested the code `__( 'ltr', { context: 'text direction' } );` locally but  for arabic it just returned `ltr` instead of `rtl` 🤔 .

I was able to inspect the contents of the file `./calypso-strings.pot` and yep, `ltr` wasn't in it.
I did find `rtl` tucked away inside `localeData[''].momentjs_locale.textDirection`, it might be possible to use that? https://github.com/Automattic/wp-calypso/pull/43132/files#diff-2d167e48666aab3480ccd61930e3cb7aR55

But hmmm I still can't understand why the original code or your suggestion is not working, do you have any more insights or could you explain again your understanding of what's going on here? 🙏 

cc @akirk I've heard you also have insights on i18n stuff ;)

Reply:
If you've added `__( 'ltr', { context: 'text direction' } );` after dev process has been started, it wouldn't work `calypso-strings.pot` and translation chunks are only being generated once on boot.

If that's the case, you can either restart the dev process, or run `yarn run build-languages` in a separate terminal window.

Reply:
ahhhh I spent some time chasing my tail on this but I now understand, just adding that line of text will cause the `calypso-i18n` tool to include a line in `./calypso-strings.pot`, which will then allow the `./bin/build-languages` script to include an entry in `ar-entry-gutenboarding.js` like so `"text direction\u0004ltr":["rtl"]`. which can then be accessed by https://github.com/WordPress/gutenberg/blob/master/packages/i18n/src/create-i18n.js#L185. 

One thing that confused me, is that using the double underscore translation method `__( 'ltr', { context: 'text direction' } );` does not actually pass the context correctly to `dcnpgettext` 🤔. the `__` method defines the second argument as `domain` https://github.com/WordPress/gutenberg/blob/master/packages/i18n/src/create-i18n.js#L118. The `_x` function does pass context as the second argument. The `_x` function is used internally by isRTL https://github.com/WordPress/gutenberg/blob/master/packages/i18n/src/create-i18n.js#L185

But when I replaced `__( 'ltr', { context: 'text direction' } );` with `_x` (`_x( 'ltr', 'text direction' );`) and ran `yarn run build-languages`, the translation for ltr was not included in the `ar-entry-gutenboarding.js` file. This would seem to be a limitation of the `i18n-calypso` cli tool which is called as part of `yarn run build-languages` and is responsible for generating the `./calypso-strings.pot` file.

Reply:
> But when I replaced __( 'ltr', { context: 'text direction' } ); with _x (_x( 'ltr', 'text direction' );) and ran yarn run build-languages, the translation for ltr was not included in the ./calypso-strings.pot or ar-entry-gutenboarding.json files. This would seem to be a limitation of the i18n-calypso cli tool which is called as part of yarn run build-languages and is responsible for generating the ./calypso-strings.pot file.

That's because `__`, `_x`, `_n`, `_nx` are used as aliases for `translate` and therefore it must be called with the same parameters format to get it extracted into `calypso-strings.pot` properly. It won't get translated properly, but since we'll only be using it to extract the string into the POT file, it should be fine for now.

We plan on switching to a different string extraction process. https://github.com/Automattic/wp-calypso/pull/39541 has been open for quite some time now, but I'll be getting back working on getting it done. It's using [wp-babel-makepot](https://github.com/Automattic/wp-babel-makepot) which is based on [babel-plugin-i18n-calypso](https://github.com/Automattic/wp-calypso/tree/master/packages/babel-plugin-i18n-calypso) and it should be handling `translate`, `__`, `_x`, `_n` and `_nx` all individually.

Reply:
If I understand everything, we need to trick the extraction into including the `ltr` with context `text direction` translation into the translation.

The correct way to call a context translation with `@wordpress/i18n` is `_x( 'ltr', 'text direction' )`, or the helper `isRTL`.

Below we still use `isRTL` to fetch the localized string. Would this creating and invoking a translation function to be picked up by the extraction have the same effect?

```ts
// We need to invoke translate in this way so that the `ltr` localization is included in the localization bundle.
const translate = () => undefined;
translate( 'ltr', { context: 'text direction' } );
```

Reply:
I haven't tested your suggestion, but it should work since it's [only matching function name](https://github.com/Automattic/xgettext-js/blob/master/xgettext.js#L214-L216).